### PR TITLE
Fix Repair-WinGetPackageManager cmdlet by retrieving dependencies from GitHub assets

### PR DESCRIPTION
**Pull request by ryfu-msft**
This change address an issue with Repair-WinGetPackageManager not installing the correct version of VCLibs before installing winget. The reason this issue came up was because the aka.ms links for the VCLibs package are deprecated and now point to an out of date package. 

The solution here is to include both a `DesktopAppInstaller_Dependencies.json` and  `DesktopAppInstaller_Dependencies.zip` asset along with each release. Repair-WinGetPackageManager will look for those files and check what dependencies are required from the json file. If any of those dependencies are missing, it will download the zip, extract, and install the required binaries based on the system architecture. 

Testing:
Verified with the fresh Windows Sandbox install and trying out the assets on an earlier prerelease.

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/winget-cli/pull/4923)

Related to:
* #4903 
* #4916

### CODE REVIEW

File: src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs [hunk 1]
```diff
@@ -63,6 +67,9 @@ internal class AppxModuleHelper
 
         // Assets
         private const string MsixBundleName = "Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle";
+        private const string DependenciesAssetName = "DesktopAppInstaller_Dependencies";
```
Comment (JohnMcPMS) [inline src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs:70]: Why is this needed?  A search on the PR didn't find it being used.

File: src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs [hunk 2]
```diff
@@ -74,7 +81,11 @@ internal class AppxModuleHelper
         private const string VCLibsUWPDesktopArm = "https://aka.ms/Microsoft.VCLibs.arm.14.00.Desktop.appx";
         private const string VCLibsUWPDesktopArm64 = "https://aka.ms/Microsoft.VCLibs.arm64.14.00.Desktop.appx";
 
+        private const string AppxPackageTemplate = "{0}_{1}_{2}__8wekyb3d8bbwe.appx";
```
Comment (JohnMcPMS) [inline src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs:84]: I don't see any reason to have the publisher hash in this string.

File: src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs [hunk 3]
```diff
@@ -74,7 +81,11 @@ internal class AppxModuleHelper
         private const string VCLibsUWPDesktopArm = "https://aka.ms/Microsoft.VCLibs.arm.14.00.Desktop.appx";
         private const string VCLibsUWPDesktopArm64 = "https://aka.ms/Microsoft.VCLibs.arm64.14.00.Desktop.appx";
 
+        private const string AppxPackageTemplate = "{0}_{1}_{2}__8wekyb3d8bbwe.appx";
+
         // Xaml
+        private const string XamlPackageName = "Microsoft.UI.Xaml.2.8";
```
Comment (JohnMcPMS) [inline src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs:87]: This is an existing string (`XamlPackage28`), why do we need another name for it?

File: src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs [hunk 4]
```diff
@@ -319,53 +330,69 @@ private async Task AddAppInstallerBundleAsync(string releaseTag, bool downgrade,
...
+        {
+            var result = this.ExecuteAppxCmdlet(
+                GetAppxPackage,
+                new Dictionary<string, object>
+                {
+                    { Name, packageName },
+                });
+
+            // See if the minimum (or greater) version is installed.
+            // TODO: Pull the minimum version from the target package
```
Comment (JohnMcPMS) [inline src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs:393]: I would remove this TODO; we are intentionally not doing that now because it is very expensive to read the manifest from a remote package as compared to the small json file we are creating.

File: src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs [hunk 5]
```diff
@@ -384,24 +411,30 @@ private async Task InstallVCLibsDependenciesAsync()
...
+                        this.pwshCmdlet.Write(StreamType.Verbose, $"{packageName} is lower than minimum required version [{minimumVersion}]: {psobject?.PackageFullName ?? "<null>"}");
                     }
                 }
             }
+        }
+
+        private async Task InstallVCLibsDependenciesFromUriAsync()
+        {
+            Dictionary<string, string> vcLibsDependencies = this.GetVCLibsDependencies();
+            this.FindMissingDependencies(vcLibsDependencies, VCLibsUWPDesktop, VCLibsUWPDesktopVersion);
```
Comment (JohnMcPMS) [inline src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs:437]: Update the value of `VCLibsUWPDesktopVersion` to be the one that 1.8 requires (33321?).

File: src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs [hunk 6]
```diff
@@ -419,6 +452,109 @@ private async Task InstallVCLibsDependenciesAsync()
...
+                                { ErrorAction, Stop },
+                                });
+                    }
+                }
+
+                return true;
+            }
+            catch (WinGetRepairException e)
+            {
+                this.pwshCmdlet.Write(StreamType.Verbose, $"Dependency assets not found in GitHub release.");
```
Comment (JohnMcPMS) [inline src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs:520]: Please add a `TryGetAsset` instead of relying on exceptions for control flow.

File: src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs [hunk 7]
```diff
@@ -419,6 +452,109 @@ private async Task InstallVCLibsDependenciesAsync()
...
 
+        private async Task<bool> TryInstallDependenciesFromGitHubArchive(string releaseTag)
+        {
+            try
+            {
+                var githubClient = new GitHubClient(RepositoryOwner.Microsoft, RepositoryName.WinGetCli);
+                var release = await githubClient.GetReleaseAsync(releaseTag);
+
+                var dependenciesJsonAsset = release.GetAsset(DependenciesJsonName);
+                var dependenciesZipAsset = release.GetAsset(DependenciesZipName);
```
Comment (JohnMcPMS) [inline src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs:463]: When this gets replaced with a `TryGetAsset`, we should still allow successful "install" of dependencies if only the json is available and all requirements are met.

File: src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs [hunk 8]
```diff
@@ -419,6 +452,109 @@ private async Task InstallVCLibsDependenciesAsync()
...
+                if (missingDependencies.Count != 0)
+                {
+                    using var dependenciesZipFile = new TempFile();
+                    using var extractedDirectory = new TempDirectory();
+                    await this.httpClientHelper.DownloadUrlWithProgressAsync(dependenciesZipAsset.BrowserDownloadUrl, dependenciesZipFile.FullPath, this.pwshCmdlet);
+                    ZipFile.ExtractToDirectory(dependenciesZipFile.FullPath, extractedDirectory.FullDirectoryPath);
+
+                    foreach (var entry in missingDependencies)
+                    {
+                        string fullPath = entry;
```
Comment (JohnMcPMS) [inline src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs:499]: This doesn't look like it will be a full path.
```suggestion
                        string fullPath = Path.Join(extractedDirectory.FullDirectoryPath, entry);
```

**Review by JohnMcPMS — COMMENTED** [review]
This design requires that both the json and archive files be present to operate.  The initial design I communicated did not.  It wouldn't take much change here to realign.

Some of my comments already indicate how to verify with only the json present, although that particular scenario isn't very useful.

What we may find more useful would be functioning (albeit inefficiently) if the json file is missing or corrupted.  That would require:
1. Placing the dependencies in the archive in architecture directories
2. Changing the format string to mirror this
3. Implementing a code path that installs all packages found in all appropriate architectures when we failed to find or parse the json

I think it at least worth doing 1 and 2, as they require very little change.  3 can be done at some future point if we want, but without having done 1 or 2 now it wouldn't work for archives we make now.

File: src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs (hunk 1 above)
Comment (ryfu-msft) [reply src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs:70]: That was meant to be the name of the extracted zip folder which I included in the asset full path.

File: src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs (hunk 3 above)
Comment (ryfu-msft) [reply src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs]: Removed

File: src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs [hunk 9]
```diff
@@ -419,6 +449,108 @@ private async Task InstallVCLibsDependenciesAsync()
...
+                    this.pwshCmdlet.Write(StreamType.Verbose, $"Dependencies zip asset not found on GitHub asset.");
+                    return false;
+                }
+
+                await this.httpClientHelper.DownloadUrlWithProgressAsync(dependenciesZipAsset.BrowserDownloadUrl, dependenciesZipFile.FullPath, this.pwshCmdlet);
+                ZipFile.ExtractToDirectory(dependenciesZipFile.FullPath, extractedDirectory.FullDirectoryPath);
+
+                foreach (var entry in missingDependencies)
+                {
+                    string fullPath = System.IO.Path.Combine(extractedDirectory.FullDirectoryPath, DependenciesAssetName, entry);
```
[the remark this answers is not in the document]
Comment (ryfu-msft) [reply src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs:506]: I'm not sure why but from my testing, extracting the zip would always create a new directory that matched the name of the GitHub asset.

[the remark this answers is not in the document]
Comment (ryfu-msft) [reply src/PowerShell/Microsoft.WinGet.Client.Engine/Helpers/AppxModuleHelper.cs]: I removed this extra directory as its not needed anymore. Basically, I compressed the folder containing the arch folders. I had to instead select all the arch folders and compress from there.

**Review by JohnMcPMS — APPROVED** [review]

**Comment by mdanish-kh** [comment]
Does this address #4778 as well?

**Comment by JohnMcPMS** [comment]
> Does this address #4778 as well?

It addresses the `Repair-WinGetPackageManager` portion of it, but we are not trying to address the broader availability through this change.

**Review by JlynnD31 — APPROVED** [review]

**Comment by sebgod** [comment]
Is it possible to generate a new preview that includes this change?

**Comment by Trenly** [comment]
@sebgod - this change should already be in the latest released version of the PowerShell module

**Comment by sebgod** [comment]
> @sebgod - this change should already be in the latest released version of the PowerShell module

Oh I confused [JlynnD31](https://github.com/JlynnD31) approved these changes [on Dec 1, 2024](https://github.com/microsoft/winget-cli/pull/4923#pullrequestreview-2471038499)
with the merge date, sorry!

Does this fix also fix this issue? https://github.com/microsoft/winget-cli/issues/5020, i.e. instead of running these instructions: https://learn.microsoft.com/en-us/windows/package-manager/winget/#install-winget-on-windows-sandbox
```
$progressPreference = 'silentlyContinue'
Write-Host "Installing WinGet PowerShell module from PSGallery..."
Install-PackageProvider -Name NuGet -Force | Out-Null
Install-Module -Name Microsoft.WinGet.Client -Force -Repository PSGallery | Out-Null
Write-Host "Using Repair-WinGetPackageManager cmdlet to bootstrap WinGet..."
Repair-WinGetPackageManager
Write-Host "Done."
```

use these instead `-IncludePrerelease`?

```
$progressPreference = 'silentlyContinue'
Write-Host "Installing WinGet PowerShell module from PSGallery..."
Install-PackageProvider -Name NuGet -Force | Out-Null
Install-Module -Name Microsoft.WinGet.Client -Force -Repository PSGallery | Out-Null
Write-Host "Using Repair-WinGetPackageManager cmdlet to bootstrap WinGet..."
Repair-WinGetPackageManager -IncludePrerelease
Write-Host "Done."
```

Or is this a separate issue?

Thanks

**Comment by denelon** [comment]
@sebgod I think the ARM issue is separate from this PR. We will most likely be cutting a new release candidate build next week. This issue was related to getting the correct UI.XAML and VCLibs dependencies installed.
